src/Couchbase.Lite.Shared/API/Sync/BackgroundMonitor.cs

```diff
@@ -0,0 +1,135 @@
+
+internal sealed class BackgroundMonitor
+{
+    private nint _bgTask = UIApplication.BackgroundTaskInvalid;
+    private NSObject? _bgObserver;
+    private NSObject? _fgObserver;
+    private readonly object _locker = new object();
+
+    public event EventHandler? OnAppBackgrounding;
+    public event EventHandler? OnAppForegrounding;
+    public event EventHandler? OnBackgroundTaskExpired;
+
+    public void Start()
+    {
+        if(IsRunningInAppExtension()) {
+            return;
+        }
+
+        _bgObserver = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.DidEnterBackgroundNotification, AppBackgrounding);
+        _fgObserver = NSNotificationCenter.DefaultCenter.AddObserver(UIApplication.WillEnterForegroundNotification, AppForegrounding);
+
+        // Already in the background? Better start a background session now:
+        DispatchQueue.MainQueue.DispatchAsync(() =>
+        {
+            if(UIApplication.SharedApplication.ApplicationState == UIApplicationState.Background) {
+                AppBackgrounding(null);
+            }
+        });
+    }
+
+    public void Stop()
+    {
+        if (IsRunningInAppExtension()) {
+            return;
+        }
+
+        if (_bgObserver != null) {
+            NSNotificationCenter.DefaultCenter.RemoveObserver(_bgObserver);
+            _bgObserver = null;
+        }
+
+        if(_fgObserver != null) {
+            NSNotificationCenter.DefaultCenter.RemoveObserver(_fgObserver);
+            _fgObserver = null;
+        }
+
+        EndBackgroundTask();
+    }
+
+    public bool EndBackgroundTask()
+    {
+        if (IsRunningInAppExtension()) {
+            return false;
+        }
+
+        lock(_locker) {
+            if(_bgTask == UIApplication.BackgroundTaskInvalid) {
+                return false;
+            }
+
+            UIApplication.SharedApplication.EndBackgroundTask(_bgTask);
+            _bgTask = UIApplication.BackgroundTaskInvalid;
+            return true;
+        }
+    }
+
+    public bool BeginBackgroundTask(string name)
+    {
+        if (IsRunningInAppExtension()) {
+            return false;
+        }
+
+        lock (_locker) {
+            if (_bgTask == UIApplication.BackgroundTaskInvalid) {
+                _bgTask = UIApplication.SharedApplication.BeginBackgroundTask(name, () =>
+                {
+                    if (_bgTask != UIApplication.BackgroundTaskInvalid) {
+                        OnBackgroundTaskExpired?.Invoke(this, EventArgs.Empty);
+                        EndBackgroundTask();
+                    }
+                });
+            }
+
+            return _bgTask != UIApplication.BackgroundTaskInvalid;
+        }
+    }
+
+    public bool HasBackgroundTask()
+    {
+        lock(_locker) {
+            return _bgTask != UIApplication.BackgroundTaskInvalid;
+        }
+    }
+
+    private static bool IsRunningInAppExtension() => Path.GetExtension(NSBundle.MainBundle.BundlePath) == "appex";
```

Comment:
Ported from Jens' extension.  An app extension on iOS is basically like a stub that doesn't run independently, but is consumed by another application even though it is installed separately.  I guess so it can be shared between multiple programs.